core/chaincode/lifecycle/lifecycle.go

```diff
@@ -674,7 +677,20 @@ func (ef *ExternalFunctions) InstallChaincode(chaincodeInstallPackage []byte) (*
 		return nil, errors.WithMessage(err, "could not save cc install package")
 	}
 
+	buildLock := ef.getBuildLock(packageID)
+	buildLock.Lock()
+	defer buildLock.Unlock()
+
 	buildStatus, ok := ef.BuildRegistry.BuildStatus(packageID)
+	if ok {
+		// another invocation of lifecycle has concurrently
+		// installed a chaincode with this package id
+		<-buildStatus.Done()
+		if buildStatus.Err() == nil {
+			return nil, errors.New("chaincode already successfully installed")
+		}
+		buildStatus, ok = ef.BuildRegistry.ResetBuildStatus(packageID)
+	}
```

Comment:
Why check ok here? Either, the previous ok was false, or, we just set ok to be false, since we didn't return, and ResetBuildStatus always returns false.

Reply:
Excellent point. I've been looking at some of this too much and it was hard to take a step back. Fixed.